### PR TITLE
Add test framework

### PR DESCRIPTION
**Pull request by geekLucian**
**Related issue**: https://github.com/chipsalliance/rocket-chip/issues/2953

**Type of change**: feature request

**Impact**: API addition (no impact on existing code)

**Development Phase**: implementation

**Release Notes**

Add test framework from @sequencer's [playground](https://github.com/sequencer/playground/tree/master/sanitytests/rocketchip).

### CODE REVIEW

File: build.sc [hunk 1]
```diff
@@ -35,6 +36,50 @@ object hardfloatRocket extends hardfloat.build.hardfloat {
...
+    //    if (!helper.isMac) {
+    //      val x86Dir = T.ctx.dest
+    //      val riscv64Dir = T.ctx.dest / "riscv64"
+    //      os.proc("make", s"DESTDIR=${x86Dir}", "install").call(spike.compile())
+    //      os.proc("make", s"DESTDIR=${riscv64Dir}", "install").call(compilerrt.compile())
+    //      os.proc("make", s"DESTDIR=${riscv64Dir}", "install").call(musl.compile())
+    //      os.copy.into(pk.compile(), riscv64Dir)
+    //    }
+    T.ctx.dest
+  }
```
Comment (sequencer) [inline build.sc:77]: Since the test used the fesvr, you may need introduce spike as well.

**Comment by michael-etzkorn** [comment]
Relevant issue:
https://github.com/chipsalliance/rocket-chip/issues/2953 

I've yet to look over the details of this but I'd like to see more `test/scala` at some point for development. 

My two cents on structuring the RTL based testing directory-wise is that it should be mergable https://github.com/chipsalliance/rocket-chip/tree/master/src/main/scala/unittest and ideally be invokable within `regression`

**Comment by sequencer** [comment]
> I don't think I understand why this is in a separate sanitytests directory instead of src/test. The changes also introduce more places where chisel/firrtl version numbers are encoded, which increases maintenance burden slightly.

I agree, I thought @geekLucian just copy-paste from my own repository, which obviously needs more refactor to make it fit into rocket-chip.

> This approach also seems to reimplement some functionality already in chiseltest. What functionality does this approach support over chiseltest?

IIRC, this approach is copy-paste from the rocket-chip original test suite, and has no relation to the chiseltest.
programs are supported to be loaded via DM, since the DMI, or JTAG-VPI depends on DPI, which makes chiseltest impossible to test it.

> Lastly, if we do move forwards with this approach, I think the csrc files should symlink the existing files, and the bootrom should use the default one in RC (Don't we want to test that that bootrom works?).

Agree, `bootrom` should use the original `bootrom` from Rocket-Chip.

File: build.sc [hunk 2]
```diff
@@ -35,6 +36,50 @@ object hardfloatRocket extends hardfloat.build.hardfloat {
   ) else Agg.empty[Dep]
 }
 
+object sanitytests extends ScalaModule with TestModule.Utest {
```
Comment (sequencer) [inline build.sc:39]: This should live under `rocketchip` project

File: build.sc [hunk 3]
```diff
@@ -35,6 +36,50 @@ object hardfloatRocket extends hardfloat.build.hardfloat {
...
+
+  override def scalacPluginClasspath = T { super.scalacPluginClasspath() }
+
+  object ivys {
+    val chisel3 = ivy"edu.berkeley.cs::chisel3:3.5.3"
+    val firrtl = ivy"edu.berkeley.cs::firrtl_2.13:1.5.3"
+    val scalatest = ivy"org.scalatest::scalatest:3.2.0"
+    val oslib = ivy"com.lihaoyi::os-lib:0.7.8"
+    val utest = ivy"com.lihaoyi::utest:0.7.10"
+  }
```
Comment (sequencer) [inline build.sc:54]: this should inherit from `rocketchip` object

File: build.sc [hunk 4]
```diff
@@ -35,6 +36,50 @@ object hardfloatRocket extends hardfloat.build.hardfloat {
...
+
+  override def ivyDeps = Agg(
+    ivys.chisel3,
+    ivys.scalatest,
+    ivys.oslib,
+    ivys.utest
+  )
```
Comment (sequencer) [inline build.sc:61]: same here.

File: build.sc [hunk 5]
```diff
@@ -35,6 +36,50 @@ object hardfloatRocket extends hardfloat.build.hardfloat {
...
+  def libraryResources = T.sources {
+    //    if (!helper.isMac) {
+    //      val x86Dir = T.ctx.dest
+    //      val riscv64Dir = T.ctx.dest / "riscv64"
+    //      os.proc("make", s"DESTDIR=${x86Dir}", "install").call(spike.compile())
+    //      os.proc("make", s"DESTDIR=${riscv64Dir}", "install").call(compilerrt.compile())
+    //      os.proc("make", s"DESTDIR=${riscv64Dir}", "install").call(musl.compile())
+    //      os.copy.into(pk.compile(), riscv64Dir)
+    //    }
+    T.ctx.dest
```
Comment (sequencer) [inline build.sc:76]: You should introduce dependency of `spike` aka riscv-isa-sim. for `csrc` you need use the original `csrc`.

File: sanitytests/src/TestConfig.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            elf,
+            bin
+          ).call()
+          proc(
+            "dd",
+            s"if=$bin",
+            s"of=$img",
+            "bs=128",
+            "count=1"
+          ).call()
```
Comment (sequencer) [inline sanitytests/src/TestConfig.scala:41]: I think compiling of `bootrom` should live outside `TestFramework`, it should somehow live in mill, which makes it being able to be cached by the build system.

File: sanitytests/src/VerilatorTest.scala [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        ).call(outputDirectory)
+        test("Hello World!") {
+          os.proc(
+            s"$emulator",
+            s"${resource("riscv64/pk")}",
+            "hello",
+          ).call(outputDirectory)
+        }
+      }
+    }
```
Comment (sequencer) [inline sanitytests/src/VerilatorTest.scala:47]: The build of test cases should also live inside the build system.
and it depends a C compiler, which should be optional as well, IMO, the RISCV toolchain is stable enough, which is possible to use the toolchain from upstream(LLVM or GCC).
You can install it in the GItHub Action like playground did.

**Review by sequencer — CHANGES_REQUESTED** [review]

**Comment by sequencer** [comment]
> I don't think I understand why this is in a separate sanitytests directory instead of `src/test`.

The `src/test` uses `ScalaTest` by default, while I use the `uTest`, I think it is reasonable to stick on `ScalaTest`, since I didn't depends on the special features of `uTest`.

As for the packaging issue. IMHO, a more fine granularity provides flexibility, the architect part can be split out from the `src/test` living in another package, like what chisel did [here](https://github.com/chipsalliance/chisel3/tree/master/integration-tests/src/test/scala).

File: build.sc [hunk 8]
```diff
@@ -14,7 +16,7 @@ object configRocket extends `api-config-chipsalliance`.`build-rules`.mill.build.
   }
 
   override def pomSettings = T {
-    rocketchip.pomSettings()
+    rocketchip.`pomSettings`()
```
Comment (sequencer) [inline build.sc:19]: why

File: build.sc [hunk 9]
```diff
@@ -47,4 +154,82 @@ object rocketchip extends common.CommonRocketChip {
...
+      ).call()
+      os.proc(
+        "dd",
+        s"if=$bin",
+        s"of=$img",
+        "bs=128",
+        "count=1"
+      ).call()
+      T.ctx.dest
+    }
```
Comment (sequencer) [inline build.sc:230]: make this as an external function, so that others can add programs as test payloads.

File: build.sc [hunk 10]
```diff
@@ -47,4 +154,82 @@ object rocketchip extends common.CommonRocketChip {
...
+      os.proc("make", s"DESTDIR=${riscv64Dir}", "install").call(musl.compile())
+      os.copy.into(pk.compile(), riscv64Dir)
+      // build hello world
+      val outputDirectory = x86Dir / "VerilatorTest"
+      os.remove.all(outputDirectory)
+      os.makeDir(outputDirectory)
+      os.proc(
+        "clang",
+        "-o", "hello",
+        s"${resource("resources/csrc/hello.c")}",
```
Comment (sequencer) [inline build.sc:184]: I personal prefer using a `testcase/hello` directory to store these tests.

File: common.sc [hunk 11]
```diff
@@ -4,8 +4,8 @@ import mill.scalalib.publish._
 import coursier.maven.MavenRepository
 
 val defaultVersions = Map(
-  "chisel3" -> "3.5.0",
-  "chisel3-plugin" -> "3.5.0"
+  "chisel3" -> "3.5.3",
+  "chisel3-plugin" -> "3.5.3"
```
Comment (sequencer) [inline common.sc:8]: You need to create another PR to bump this.

**Review by sequencer — CHANGES_REQUESTED** [review]
You are going to the right direction!

**Pull request closed without merging**